File: 14-longest-common-prefix/longest-common-prefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        int p = 0;
        boolean flag = false;
        while (true) {
            char c = ' ';
            if (p < strs[0].length()) {
                c = strs[0].charAt(p);
            } else {
                return strs[0].substring(0, p);
            }
            for (int i = 1; i < strs.length; i++) {
                if (p >= strs[i].length() ||  strs[i].charAt(p) != c) {
                    flag = true;
                    break;
                }
            }
            if (!flag) {
                p++;
            } else {
                return strs[0].substring(0, p);
            }
        }
    }
}

File: 14-longest-common-prefix/longest-common-prefix.py
class TrieNode():
    def __init__(self):
        self.cnt = 0
        self.children = [None] * 26
    
class Trie():
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word):
        word_str = str(word)
        node = self.root
        for w in word_str:
            idx = ord(w) - ord('a')
            if not node.children[idx]:
                node.children[idx] = TrieNode()
            node.children[idx].cnt += 1
            node = node.children[idx]

        
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        trie = Trie()
        for s in strs:
            trie.insert(s)
        node = trie.root
        res = ""
        while node:
            last_node = node
            for i in range(26):
                if node.children[i] and node.children[i].cnt == len(strs):
                    res += chr(ord('a') + i)
                    node = node.children[i]
                    break
            if node == last_node:
                break
        return res
